src/lib/libc/locale/setlocale.3
.\"	$MirOS$
.\"	$OpenBSD: setlocale.3,v 1.11 2005/08/12 18:31:43 jmc Exp $
.\"	$NetBSD: setlocale.3,v 1.3 1997/07/14 23:19:47 kleink Exp $
.\"
.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Donn Seeley at BSDI.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)setlocale.3	8.1 (Berkeley) 6/9/93
.\"
.Dd $Mdocdate$
.Dt SETLOCALE 3
.Os
.Sh NAME
.Nm setlocale ,
.Nm localeconv
.Nd natural language formatting for C
.Sh SYNOPSIS
.Fd #include <locale.h>
.Ft char *
.Fn setlocale "int category" "const char *locale"
.Ft struct lconv *
.Fn localeconv "void"
.Sh DESCRIPTION
The
.Fn setlocale
function sets the C library's notion
of natural language formatting style
for particular sets of routines.
Each such style is called a
.Dq locale
and is invoked using an appropriate name passed as a C string.
The
.Fn localeconv
routine returns the current locale's parameters
for formatting numbers.
.Pp
The
.Fn setlocale
function recognises several categories of routines.
These are the categories and the sets of routines they select:
.Bl -tag -width LC_MONETARY
.It Dv LC_ALL
Set the entire locale generically.
.It Dv LC_COLLATE
Set a locale for string collation routines.
This controls alphabetic ordering in
.Fn strcoll
and
.Fn strxfrm .
.It Dv LC_CTYPE
Set a locale for the
.Xr ctype 3
functions.
This controls recognition of upper and lower case,
alphabetic or non-alphabetic characters, and so on.
.Pp
In the current implementation, one additional locale
is supported in this category, namely
.Dv en_US.UTF-8 ,
which is selected if the locale string matches
.Dq UTF-8
or
.Dq utf8 ,
case-insensitively.
.It Dv LC_MONETARY
Set a locale for formatting monetary values;
this affects the
.Fn localeconv
function.
.It Dv LC_NUMERIC
Set a locale for formatting numbers.
This controls the formatting of decimal points
in input and output of floating point numbers
in functions such as
.Fn printf
and
.Fn scanf ,
as well as values returned by
.Fn localeconv .
.It Dv LC_TIME
Set a locale for formatting dates and times using the
.Fn strftime
function.
.El
.Pp
Only three locales are defined by default,
the empty string
.Li "\&""\|""
which denotes the native environment, and the
.Li "\&""C""
and
.Li "\&""POSIX""
locales, which denote the C language environment.
A
.Fa locale
argument of
.Dv NULL
causes
.Fn setlocale
to return the current locale.
By default, C programs start in the
.Li "\&""C""
locale.
The only function in the library that sets the locale is
.Fn setlocale ;
the locale is never changed as a side effect of some other routine.
.Pp
The
.Fn localeconv
function returns a pointer to a structure
which provides parameters for formatting numbers,
especially currency values:
.Bd -literal -offset indent
struct lconv {
	char	*decimal_point;
	char	*thousands_sep;
	char	*grouping;
	char	*int_curr_symbol;
	char	*currency_symbol;
	char	*mon_decimal_point;
	char	*mon_thousands_sep;
	char	*mon_grouping;
	char	*positive_sign;
	char	*negative_sign;
	char	int_frac_digits;
	char	frac_digits;
	char	p_cs_precedes;
	char	p_sep_by_space;
	char	n_cs_precedes;
	char	n_sep_by_space;
	char	p_sign_posn;
	char	n_sign_posn;
};
.Ed
.Pp
The individual fields have the following meanings:
.Bl -tag -width mon_decimal_point
.It Fa decimal_point
The decimal point character, except for currency values.
.It Fa thousands_sep
The separator between groups of digits
before the decimal point, except for currency values.
.It Fa grouping
The sizes of the groups of digits, except for currency values.
This is a pointer to a vector of integers, each of size
.Li char ,
representing group size from low order digit groups
to high order (right to left).
The list may be terminated with 0 or
.Dv CHAR_MAX .
If the list is terminated with 0,
the last group size before the 0 is repeated to account for all the digits.
If the list is terminated with
.Dv CHAR_MAX ,
no more grouping is performed.
.It Fa int_curr_symbol
The standardised international currency symbol.
.It Fa currency_symbol
The local currency symbol.
.It Fa mon_decimal_point
The decimal point character for currency values.
.It Fa mon_thousands_sep
The separator for digit groups in currency values.
.It Fa mon_grouping
Like
.Fa grouping
but for currency values.
.It Fa positive_sign
The character used to denote non-negative currency values,
usually the empty string.
.It Fa negative_sign
The character used to denote negative currency values,
usually a minus sign.
.It Fa int_frac_digits
The number of digits after the decimal point
in an international-style currency value.
.It Fa frac_digits
The number of digits after the decimal point
in the local style for currency values.
.It Fa p_cs_precedes
1 if the currency symbol precedes the currency value
for non-negative values, 0 if it follows.
.It Fa p_sep_by_space
1 if a space is inserted between the currency symbol
and the currency value for non-negative values, 0 otherwise.
.It Fa n_cs_precedes
Like
.Fa p_cs_precedes
but for negative values.
.It Fa n_sep_by_space
Like
.Fa p_sep_by_space
but for negative values.
.It Fa p_sign_posn
The location of the
.Fa positive_sign
with respect to a non-negative quantity and the
.Fa currency_symbol ,
coded as follows:
.Pp
.Bl -tag -width 3n -compact
.It Li 0
Parentheses around the entire string.
.It Li 1
Before the string.
.It Li 2
After the string.
.It Li 3
Just before
.Fa currency_symbol .
.It Li 4
Just after
.Fa currency_symbol .
.El
.It Fa n_sign_posn
Like
.Fa p_sign_posn
but for negative currency values.
.El
.Pp
Unless mentioned above,
an empty string as a value for a field
indicates a zero length result or
a value that is not in the current locale.
A
.Dv CHAR_MAX
result similarly denotes an unavailable value.
.Sh RETURN VALUES
The
.Fn setlocale
function returns
.Dv NULL
and fails to change the locale
if the given combination of
.Fa category
and
.Fa locale
makes no sense.
The
.Fn localeconv
function returns a pointer to a static object
which may be altered by later calls to
.Fn setlocale
or
.Fn localeconv .
.Sh STANDARDS
The
.Fn setlocale
and
.Fn localeconv
functions conform to
.St -ansiC .
.Sh HISTORY
The
.Fn setlocale
and
.Fn localeconv
functions first appeared in
.Bx 4.4 .
.Sh BUGS
The current implementation supports only the
.Li "\&""C""
and
.Li "\&""POSIX""
locales for all but the
.Dv LC_CTYPE
locale,
and only the
.Li "\&""en_US.UTF-8""
.Dv LC_CTYPE
locale.
Soon, to remove this bug, the
.Li "\&""C.UTF-8"",
.Li "\&""C""
and
.Li "\&""POSIX""
locales will be supported for
.Dv LC_CTYPE .
.Pp
In spite of the gnarly currency support in
.Fn localeconv ,
the standards don't include any functions
for generalised currency formatting.
.Pp
.Dv LC_COLLATE
does not make sense for many languages.
Use of
.Dv LC_MONETARY
could lead to misleading results until we have a real time currency
conversion function.
.Dv LC_NUMERIC
and
.Dv LC_TIME
are personal choices and should not be wrapped up with the other categories.


contrib/code/Snippets/setlocale.c
/*-
 * Copyright (c) 2005, 2007
 *	Thorsten Glaser <tg@mirbsd.de>
 *
 * Provided that these terms and disclaimer and all copyright notices
 * are retained or reproduced in an accompanying document, permission
 * is granted to deal in this work without restriction, including un-
 * limited rights to use, publicly perform, distribute, sell, modify,
 * merge, give away, or sublicence.
 *
 * This work is provided "AS IS" and WITHOUT WARRANTY of any kind, to
 * the utmost extent permitted by applicable law, neither express nor
 * implied; without malicious intent or gross negligence. In no event
 * may a licensor, author or contributor be held liable for indirect,
 * direct, other damage, loss, or other issues arising in any way out
 * of dealing in the work, even if advised of the possibility of such
 * damage or existence of a defect, except proven that it results out
 * of said person's immediate fault when using the work as intended.
 */

#define setlocale donotdeclareme

#include <sys/types.h>
#include <langinfo.h>
#include <locale.h>
#include <nl_types.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

#undef setlocale

static const char __rcsid_1[] = "$miros: src/lib/libc/i18n/charsets.c,v 1.17 2007/06/07 16:41:04 tg Exp $";
static const char __rcsid_2[] = "$MirOS: contrib/code/Snippets/setlocale.c,v 1.4 2007/06/07 16:41:04 tg Exp $";

/* simple locale support (C and UTF-8) */

char *setlocale(int, const char *);

static bool locale_is_utf8 = false;

#define STR_C	"C"
#define STR_UTF	"en_US.UTF-8"

char *
setlocale(int category, const char *locale)
{
	static char lc_other[] = STR_C;
	static char lc_ctype[] = STR_UTF;

	if ((category != LC_CTYPE) && (category != LC_ALL)) {
		memmove(lc_other, STR_C, sizeof (lc_other));
		return (lc_other);
	}

	if (locale == NULL)
		goto getlocale;

	if (*locale == '\0') {
		locale = getenv("LC_ALL");
		if (!locale || !*locale) {
			locale = getenv("LC_CTYPE");
			if (!locale || !*locale)
				locale = getenv("LANG");
			if (!locale)
				locale = "C";
		}
	}

	locale_is_utf8 = (strcasestr(locale, "UTF-8")
	    || strcasestr(locale, "utf8"));

 getlocale:
	strlcpy(lc_ctype, locale_is_utf8 ? STR_UTF : STR_C, sizeof (lc_ctype));
	return (lc_ctype);
}
